[WIP] builder-worker: kubernetes exporter integration

robertgzr opened the pull request:
~~This is mostly for a demo. It has hardcoded paths and configs!~~

I cleaned up the implementation, removed the hard-coded values and replaced them with API integrations (like the docker exporter does).
There is no UI for this integration yet but it should be possible via raw API calls.

TODO:
- [x] use kubectl from hab pkg
- [ ] fix the missing parts https://github.com/kinvolk/habitat/pull/8#issuecomment-351742190 
- [ ] a way to set the integration options via UI (not this PR)

---

to test these changes:

0. do `git cherry-pick origin/master..kinvolk/schu/vagrant`
1. setup local builder. Follow the steps [here](https://github.com/kinvolk/habitat/blob/schu/vagrant/BUILDER_VAGRANT.md)
2. we also need `core/kubernetes` in the local core origin:
```bash
$ hab pkg install core/kubernetes
$ package_load /hab/cache/artifacts/core-kubernetes-*.hart
```
3. since there is no ui yet we need to create the integrations from the cli
```bash
$ http POST localhost:9636/v1/depot/origins/<your_origin>/integrations/kubernetes/<integration_name> "Authorization:Bearer:${HAB_ORIGIN_AUTH}" kubeconfig_path=<path_to_kubeconfig>
```
4. upload the kubeconfig to the path set above

---

So far I have not managed to test the full integration myself

robertgzr committed:
builder-worker: add kubernetes exporter

Adds kubernetes-exporter integration to the worker.
Modeled after the docker integration.

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

robertgzr committed:
builder-worker: add kubernetes-exporter spec

using hardcoded values

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

zeenix commented on components/builder-worker/src/runner/mod.rs:
```diff
@@ -504,6 +520,11 @@ impl Runner {
     }
+
+    fn has_kubernetes_integration(&self) -> bool {
+        // demo
```
this should be a FIXME perhaps so it stands out? :)

zeenix commented on components/builder-worker/src/runner/kubernetes.rs:
```diff
@@ -26,10 +26,11 @@ use runner::workspace::Workspace;
-        include_str!(concat!(env!("OUT_DIR"), "/KUBERNETES_EXPORTER_PKG_IDENT")),
-    );
+    static ref KUBERNETES_EXPORTER_PROGRAM: PathBuf = PathBuf::from("/src/target/debug/hab-pkg-export-kubernetes");
+    // static ref KUBERNETES_EXPORTER_PROGRAM: PathBuf = hfs::resolve_cmd_in_pkg(
```
A comment above here explaining why this is commented-out would be nice.

zeenix commented:
Apart from the comments, looks fine to my naive eyes. Oh and before you submit to upstream, don't forget to sign your patches, as in `signed-off-by` tags.

robertgzr committed:
builder-worker: collect stderr with log_pipe

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

robertgzr committed:
builder-worker: move spec code to runner::util

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

robertgzr committed:
builder-worker: set kubectl path using env var

To make it similar to DOCKER_HOST

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

robertgzr committed:
builder-worker: support multiple integration types in runner::util

Until now the validation and generating the spec assumed only one
element, docker integration. This commit filters both for integration
type.

The validation function was split to allow handling the checks per
integration type.

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

robertgzr committed:
builder-worker: log stderr to file

Also reverts commit 441b856ed1482ea02c7c8f9eea80f5f592bd0db2
"builder-worker: collect stderr with log_pipe"

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

robertgzr committed:
builder-worker: use kubectl from habitat

Looks in the habitat pkg KUBECTL_PKG_IDENT for the binary

Signed-off-by: Robert Günzler <robertguenzler@kinvolk.io>

robertgzr commented on components/builder-worker/build.rs:
```diff
@@ -39,3 +41,23 @@ fn write_docker_pkg_ident() {
+        // Use the value provided by the build system if present
+        Ok(ident) => ident,
+        // Use the latest installed package as a default for development
+        _ => String::from("core/kubernetes"),
```
I use `core/kubernetes` atm @schu 
Will that work?

schu replied:
I should work as 1.8.3 has been promoted to stable (https://bldr.habitat.sh/#/pkgs/core/kubernetes) and 1.7.5 was there already.

I've also submitted a [kubectl plan](https://github.com/habitat-sh/core-plans/pull/1058), pending.

robertgzr replied:
Thanks 👍

schu replied:
`core/kubectl` is now available.

robertgzr commented:
This needs more work:
<details>

```
worker.1    | thread 'job_runner' panicked at 'Project integrations must not be empty', /checkout/src/libcore/option.rs:839:4
worker.1    | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
worker.1    | stack backtrace:
worker.1    |    0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
worker.1    |              at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
worker.1    |    1: std::sys_common::backtrace::_print
worker.1    |              at /checkout/src/libstd/sys_common/backtrace.rs:69
worker.1    |    2: std::panicking::default_hook::{{closure}}
worker.1    |              at /checkout/src/libstd/sys_common/backtrace.rs:58
worker.1    |              at /checkout/src/libstd/panicking.rs:381
worker.1    |    3: std::panicking::default_hook
worker.1    |              at /checkout/src/libstd/panicking.rs:397
worker.1    |    4: std::panicking::rust_panic_with_hook
worker.1    |              at /checkout/src/libstd/panicking.rs:577
worker.1    |    5: std::panicking::begin_panic
worker.1    |              at /checkout/src/libstd/panicking.rs:538
worker.1    |    6: std::panicking::begin_panic_fmt
worker.1    |              at /checkout/src/libstd/panicking.rs:522
worker.1    |    7: rust_begin_unwind
worker.1    |              at /checkout/src/libstd/panicking.rs:498
worker.1    |    8: core::panicking::panic_fmt
worker.1    |              at /checkout/src/libcore/panicking.rs:71
worker.1    |    9: core::option::expect_failed
worker.1    |              at /checkout/src/libcore/option.rs:839
worker.1    |   10: <core::option::Option<T>>::expect
worker.1    |              at /checkout/src/libcore/option.rs:302
worker.1    |   11: habitat_builder_worker::runner::util::docker_exporter_spec
worker.1    |              at src/runner/util.rs:323
worker.1    |   12: habitat_builder_worker::runner::Runner::build
worker.1    |              at src/runner/mod.rs:396
worker.1    |   13: habitat_builder_worker::runner::Runner::do_build
worker.1    |              at src/runner/mod.rs:238
worker.1    |   14: habitat_builder_worker::runner::Runner::run
worker.1    |              at src/runner/mod.rs:310
worker.1    |   15: habitat_builder_worker::runner::RunnerMgr::spawn_job::{{closure}}
worker.1    |              at src/runner/mod.rs:691
```

</details>

---

I logged the integration structures that it fails on and it looks like there is some information missing...

For the origin we have the integration type available:
```rust
debug!("origin_integrations: {:?}", workspace.job.get_integrations());
===
habitat_builder_worker::runner::util: origin_integrations: [
origin: "robertgzr" integration: "docker" name: "dockerhub" body: "{\"username\":\"robertgzr\",\"password\":\"[secret]\"}",
origin: "robertgzr" integration: "kubernetes" name: "kubespawn" body: "{\"kubeconfig_path\": \"/opt/kubeconfig\"}"
]
```

but project integration doesn't seem to have it:
```rust
debug!("proj_integrations: {:?}", workspace.job.get_project_integrations());
===
habitat_builder_worker::runner::util: proj_integrations: [
origin: "robertgzr" body: "{\"docker_hub_repo_name\":\"robertgzr/it-works\",\"latest_tag\":true,\"version_tag\":true,\"version_release_tag\":true}", 
origin: "robertgzr" body: "{\"replicas\": \"1\"}"
]
```

robertgzr committed:
only lookup k8s data in origin integration

Since project integrations can not be filtered by integration type
we have to store all the data in the origin integration

robertgzr commented:
Commit https://github.com/kinvolk/habitat/pull/8/commits/cb24e53a713ae43a3bfad96e0f09286668dc4c3c should remove the things that break above.

After that the origin integration needs to be set up like:```
http PUT :9636/v1/depot/origins/robertgzr/integrations/kubernetes/kube kubeconfig_path="/opt/kubeconfig" replicas=1 Authorization:Bearer:$HAB_AUTH_TOKEN
```

I was unable to test it so far

schu commented:
AFAICT, the comand to setup origin integration [above](https://github.com/kinvolk/habitat/pull/8#issuecomment-351771837) is not correct and results in an error:

```
worker.1    | DEBUG:habitat_builder_worker::runner: Failed to validate integrations for schu/zero-dep, err=InvalidIntegrations("project integration replicas value must be an i64")
```

Instead, `--json` must be used with `:=` to make sure the replica value is actually send and stored as integer and not string.

```
http --json PUT :9636/v1/depot/origins/indradhanush/integrations/kubernetes/kube5 kubeconfig_path="/opt/kubeconfig" replicas:=1 Authorization:Bearer:$HAB_AUTH_TOKEN
```

schu commented:
This now leads to

```
worker.1    | DEBUG:habitat_builder_worker::runner::docker: completed docker export command, status=ExitStatus(ExitStatus(0))
worker.1    | DEBUG:habitat_builder_worker::runner::docker: signaling dockerd to shutdown pid=18732, sig=TERM
worker.1    | DEBUG:habitat_builder_worker::runner::docker: terminated dockerd                                   
worker.1    | DEBUG:habitat_builder_worker::runner: Found runnable package, running kubernetes export    
worker.1    | thread 'job_runner' panicked at 'Package installation for 'core/hab-pkg-export-kubernetes' not found on disk! This is required for the proper operation of this program (Err: PackageNotFound(PackageIdent { origin: "core", name: "hab-pkg-export-kubernetes", version: None, release: None }))', /src/components/core/src/fs.rs:338:12
worker.1    | note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.                              
worker.1    | stack backtrace:                                                                          
worker.1    |    0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace                                            
worker.1    |              at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49                                         
worker.1    |    1: std::sys_common::backtrace::_print                                                                  
worker.1    |              at /checkout/src/libstd/sys_common/backtrace.rs:69                                                     
worker.1    |    2: std::panicking::default_hook::{{closure}}                                           
worker.1    |              at /checkout/src/libstd/sys_common/backtrace.rs:58                           
worker.1    |              at /checkout/src/libstd/panicking.rs:381                                                     
worker.1    |    3: std::panicking::default_hook                                                                                  
worker.1    |              at /checkout/src/libstd/panicking.rs:397                                                     
worker.1    |    4: std::panicking::rust_panic_with_hook                                                                          
worker.1    |              at /checkout/src/libstd/panicking.rs:577                                     
worker.1    |    5: std::panicking::begin_panic                                                                         
worker.1    |              at /checkout/src/libstd/panicking.rs:538                                                               
worker.1    |    6: habitat_core::fs::resolve_cmd_in_pkg                                                
worker.1    |              at ./components/core/src/fs.rs:338                                           
worker.1    |    7: <habitat_builder_worker::runner::kubernetes::KUBERNETES_EXPORTER_PROGRAM as core::ops::deref::Deref>::deref::__static_ref_initialize
worker.1    |              at src/runner/kubernetes.rs:30                                          
worker.1    |    8: core::ops::function::FnOnce::call_once                                                              
worker.1    |              at /checkout/src/libcore/ops/function.rs:223                                                           
worker.1    |    9: <lazy_static::lazy::Lazy<T>>::get::{{closure}}                                      
worker.1    |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/lazy_static-0.2.8/src/lazy.rs:23
worker.1    |   10: std::sync::once::Once::call_once::{{closure}}                                       
worker.1    |              at /checkout/src/libstd/sync/once.rs:228                                                     
worker.1    |   11: std::sync::once::Once::call_inner                                                                             
worker.1    |              at /checkout/src/libstd/sync/once.rs:308                                     
worker.1    |   12: std::sync::once::Once::call_once                                                    
worker.1    |              at /checkout/src/libstd/sync/once.rs:228                                     
worker.1    |   13: <habitat_builder_worker::runner::kubernetes::KUBERNETES_EXPORTER_PROGRAM as core::ops::deref::Deref>::deref
worker.1    |              at /root/.cargo/registry/src/github.com-1ecc6299db9ec823/lazy_static-0.2.8/src/lazy.rs:22
worker.1    |              at ./components/builder-worker/<__lazy_static_internal macros>:20
worker.1    |              at ./components/builder-worker/<__lazy_static_internal macros>:21
worker.1    |   14: habitat_builder_worker::runner::kubernetes::KubernetesExporter::spawn_exporter
worker.1    |              at src/runner/kubernetes.rs:86
worker.1    |   15: habitat_builder_worker::runner::kubernetes::KubernetesExporter::export
worker.1    |              at src/runner/kubernetes.rs:74
worker.1    |   16: habitat_builder_worker::runner::Runner::build
worker.1    |              at src/runner/mod.rs:412
worker.1    |   17: habitat_builder_worker::runner::Runner::do_build
worker.1    |              at src/runner/mod.rs:238
worker.1    |   18: habitat_builder_worker::runner::Runner::run
worker.1    |              at src/runner/mod.rs:310
worker.1    |   19: habitat_builder_worker::runner::RunnerMgr::spawn_job::{{closure}}
worker.1    |              at src/runner/mod.rs:691
worker.1    | DEBUG:habitat_builder_worker::runner::log_pipe: completing log_forwarder, job_id=875188910966185984
```

schu commented:
After

```
hab install core/hab-pkg-export-kubernetes --channel unstable
hab install core/kubectl
```

we get past the exception above.

asymmetric commented:
@schu is step 0 `git cherry-pick a9032612fe636ebac7d5fc98085ed56253393ae1..schu/vagrant` still necessary?

indradhanush commented:
@asymmetric Yes, it is but the commit id might not be accurate. I run a `git merge origin/schu/vagrant` after running a `git fetch origin` instead.

schu commented:
Please don't merge but rather cherry-pick the range of commits (currently 6). With merge other changes are potentially applied as well.

lilic commented on components/builder-worker/src/runner/util.rs:
```diff
@@ -165,10 +186,131 @@ pub fn validate_integrations(workspace: &Workspace) -> Result<()> {
 }
 
+fn validate_kubernetes_integrations(workspace: &Workspace) -> Result<()> {
+    // Validate project integration
```
Why is this commented out?

lilic commented on components/builder-worker/src/runner/util.rs:
```diff
@@ -237,6 +385,46 @@ pub fn docker_exporter_spec(workspace: &Workspace) -> DockerExporterSpec {
+    ).expect("Origin integration body must be JSON");
+
+
+    // let opts: JsonValue = serde_json::from_str(
```
How come it's commented out?

asymmetric commented:
What's the status of this? Are we still planning on working on it?

schu commented:
We are not working on this any longer.
